File: Sources/MonitorCore/Domain.swift
import Foundation

public struct MonitorError: LocalizedError {
    public let message: String
    public let status: Int?
    public init(_ message: String, status: Int? = nil) { self.message = message; self.status = status }
    public var errorDescription: String? { message }
}

public struct Configuration: Codable, Equatable {
    public var baseURL = ""
    public var name = "CPAMP"
    public var interval = 300
    public var warning = 20
    public var critical = 10
    public var notifications = true
    public var excluded: Set<String> = []
    public var accountOrder: [String] = []
    public var sshHost = ""
    public var sshPort = 18317
    public var localPort = 18318
    public var credentialID: String { sshHost.isEmpty ? baseURL : "ssh://\(sshHost):\(sshPort)" }
    public init() {}

    private enum CodingKeys: String, CodingKey {
        case baseURL, name, interval, warning, critical, notifications, excluded, accountOrder
        case sshHost, sshPort, localPort
    }

    // accountOrder was added after the first released config format. Decode
    // it optionally so existing settings continue to load unchanged.
    public init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        baseURL = try values.decodeIfPresent(String.self, forKey: .baseURL) ?? ""
        name = try values.decodeIfPresent(String.self, forKey: .name) ?? "CPAMP"
        interval = try values.decodeIfPresent(Int.self, forKey: .interval) ?? 300
        warning = try values.decodeIfPresent(Int.self, forKey: .warning) ?? 20
        critical = try values.decodeIfPresent(Int.self, forKey: .critical) ?? 10
        notifications = try values.decodeIfPresent(Bool.self, forKey: .notifications) ?? true
        excluded = try values.decodeIfPresent(Set<String>.self, forKey: .excluded) ?? []
        accountOrder = try values.decodeIfPresent([String].self, forKey: .accountOrder) ?? []
        sshHost = try values.decodeIfPresent(String.self, forKey: .sshHost) ?? ""
        sshPort = try values.decodeIfPresent(Int.self, forKey: .sshPort) ?? 18317
        localPort = try values.decodeIfPresent(Int.self, forKey: .localPort) ?? 18318
    }

    public static func normalizeURL(_ input: String) throws -> URL {
        guard var c = URLComponents(string: input.trimmingCharacters(in: .whitespacesAndNewlines)),
              let host = c.host, !host.isEmpty, ["https", "http"].contains(c.scheme?.lowercased() ?? ""),
              c.user == nil, c.password == nil, c.query == nil, c.fragment == nil else {
            throw MonitorError("请输入完整服务地址，不包含密码、查询参数或片段。")
        }
        if c.scheme == "http" && !["localhost", "127.0.0.1", "::1"].contains(host) {
            throw MonitorError("远程服务必须使用 HTTPS，避免管理密钥明文传输。")
        }
        var path = c.path
        while path.hasSuffix("/") { path.removeLast() }
        for suffix in ["/management.html", "/v0/management", "/v1"] where path.hasSuffix(suffix) {
            path.removeLast(suffix.count)
            break
        }
        c.path = path
        c.scheme = c.scheme?.lowercased()
        guard let url = c.url else { throw MonitorError("服务地址无效。") }
        return url
    }

    public func validated() throws -> Configuration {
        var result = self
        if !sshHost.isEmpty {
            guard !sshHost.hasPrefix("-"), sshHost.range(of: "^[A-Za-z0-9_@.-]+$", options: .regularExpression) != nil,
                  (1...65535).contains(sshPort), (1024...65535).contains(localPort) else {
                throw MonitorError("SSH 主机别名或端口无效。")
            }
            result.baseURL = "http://127.0.0.1:\(localPort)"
        } else { result.baseURL = try Self.normalizeURL(baseURL).absoluteString }
        guard (60...3600).contains(interval), critical >= 1, warning <= 99, critical < warning else {
            throw MonitorError("刷新间隔需为 1–60 分钟，阈值需满足 1 ≤ 紧急 < 提醒 ≤ 99。")
        }
        return result
    }
}

public struct Account: Identifiable {
    public var id: String { provider + ":" + filename + ":" + authIndex }
    public let filename: String
    public let provider: String
    public let authIndex: String
    public let email: String
    public let status: String
    public let disabled: Bool
    public let projectID: String
    public let accountID: String
    public let plan: String
    public var title: String { email.isEmpty ? filename : email }
    public var providerName: String {
        ["codex": "Codex", "antigravity": "Antigravity", "xai": "Grok / xAI", "claude": "Claude"][provider] ?? provider
    }
    public init(json: [String: Any]) {
        filename = string(json["name"])
        provider = string(json["provider"] ?? json["type"]).lowercased()
        authIndex = string(json["auth_index"])
        email = string(json["email"])
        status = string(json["status"])
        disabled = json["disabled"] as? Bool ?? false
        projectID = string(json["project_id"])
        let account = json["account"] as? [String: Any] ?? [:]
        accountID = string(account["account_id"] ?? json["account_id"])
        plan = string(account["plan_type"] ?? json["account_type"])
    }
    public var target: [String: Any] {
        var result: [String: Any] = ["row_key": id, "auth_file_snapshot": filename,
                                    "auth_provider_snapshot": provider, "auth_index": authIndex]
        if !accountID.isEmpty { result["auth_account_id_snapshot"] = accountID }
        if !projectID.isEmpty { result["auth_project_id_snapshot"] = projectID }
        return result
    }
}

public struct QuotaWindow: Identifiable, Codable, Equatable {
    public var id: String
    public var title: String
    public var remaining: Double?
    public var reset: Date?
    public var observed: Date
    public init(id: String, title: String, remaining: Double?, reset: Date?, observed: Date) {
        self.id = id; self.title = title; self.remaining = remaining; self.reset = reset; self.observed = observed
    }
    public func fresh(now: Date, maxAge: TimeInterval) -> Bool {
        now.timeIntervalSince(observed) >= -60 && now.timeIntervalSince(observed) <= maxAge && (reset == nil || reset! > now)
    }
}

public struct QuotaResult {
    public var windows: [QuotaWindow]
    public var plan: String
    public init(windows: [QuotaWindow], plan: String = "") { self.windows = windows; self.plan = plan }
}

public func string(_ value: Any?) -> String {
    if let s = value as? String { return s }
    if let n = value as? NSNumber { return n.stringValue }
    return ""
}

public func number(_ value: Any?) -> Double? {
    guard let value, !(value is NSNull), CFGetTypeID(value as CFTypeRef) != CFBooleanGetTypeID() else { return nil }
    let n = (value as? NSNumber)?.doubleValue ?? Double(string(value))
    return n.flatMap { $0.isFinite ? $0 : nil }
}

public func timestamp(_ value: Any?) -> Date? {
    if let n = number(value), n > 0, n < 100_000_000_000_000 {
        return Date(timeIntervalSince1970: n > 100_000_000_000 ? n / 1000 : n)
    }
    let s = string(value)
    let formatter = ISO8601DateFormatter()
    formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
    if let d = formatter.date(from: s) { return d }
    formatter.formatOptions = [.withInternetDateTime]
    return formatter.date(from: s)
}

import CoreFoundation

public enum QuotaParser {
    static func percent(_ value: Any?) -> Double? {
        number(value).flatMap { (0...100).contains($0) ? $0 : nil }
    }
    public static func parse(provider: String, payload: [String: Any], now: Date = Date()) throws -> QuotaResult {
        var windows: [QuotaWindow] = []
        var plan = ""
        switch provider {
        case "codex":
            plan = string(payload["plan_type"])
            var limits: [(String, String, [String: Any])] = []
            if let main = payload["rate_limit"] as? [String: Any] { limits.append(("main", "", main)) }
            if let review = payload["code_review_rate_limit"] as? [String: Any] { limits.append(("review", "代码审查 ", review)) }
            for (index, extra) in (payload["additional_rate_limits"] as? [[String: Any]] ?? []).enumerated() {
                if let rate = extra["rate_limit"] as? [String: Any] {
                    let name = string(extra["limit_name"] ?? extra["metered_feature"])
                    limits.append(("extra-" + (name.isEmpty ? String(index) : name), name + " ", rate))
                }
            }
            for (id, prefix, limit) in limits {
                for key in ["primary_window", "secondary_window"] {
                    guard let w = limit[key] as? [String: Any] else { continue }
                    let duration = number(w["limit_window_seconds"])
                    let label = duration == 604800 ? "周额度" : duration.map { "\(Int($0 / 3600)) 小时" } ?? "额度"
                    let relative = number(w["reset_after_seconds"]).flatMap { $0 > 0 ? now.addingTimeInterval($0) : nil }
                    windows.append(.init(id: id + ":" + key, title: prefix + label,
                                         remaining: percent(w["used_percent"]).map { 100 - $0 },
                                         reset: timestamp(w["reset_at"]) ?? relative, observed: now))
                }
            }
        case "antigravity":
            for (index, group) in (payload["groups"] as? [[String: Any]] ?? []).enumerated() {
                let label = string(group["displayName"] ?? group["display_name"])
                for (bi, bucket) in (group["buckets"] as? [[String: Any]] ?? []).enumerated() {
                    let window = string(bucket["window"])
                    let suffix = window == "weekly" ? "周额度" : window
                    let fraction = number(bucket["remainingFraction"] ?? bucket["remaining_fraction"])
                    let remaining = fraction.flatMap { (0...1).contains($0) ? $0 * 100 : nil }
                    let bucketID = string(bucket["bucketId"] ?? bucket["bucket_id"])
                    windows.append(.init(id: bucketID.isEmpty ? "\(label)-\(index)-\(window)-\(bi)" : bucketID,
                                         title: "\(label.isEmpty ? "额度组 \(index + 1)" : label) \(suffix)",
                                         remaining: remaining, reset: timestamp(bucket["resetTime"] ?? bucket["reset_time"]), observed: now))
                }
            }
        case "xai":
            let config = payload["config"] as? [String: Any] ?? [:]
            let period = config["currentPeriod"] as? [String: Any] ?? [:]
            let reset = timestamp(period["end"] ?? config["billingPeriodEnd"])
            let isWeekly = string(period["type"]).contains("WEEKLY")
            if let used = percent(config["creditUsagePercent"]) {
                windows.append(.init(id: "billing", title: isWeekly ? "周额度" : "账期额度", remaining: 100 - used, reset: reset, observed: now))
            }
        case "claude":
            for key in payload.keys.sorted() where key.hasPrefix("five_hour") || key.hasPrefix("seven_day") {
                guard let w = payload[key] as? [String: Any] else { continue }
                windows.append(.init(id: key, title: key == "five_hour" ? "5 小时" : key == "seven_day" ? "周额度" : key,
                                     remaining: percent(w["utilization"]).map { 100 - $0 }, reset: timestamp(w["resets_at"]), observed: now))
            }
        default: throw MonitorError("暂不支持此提供商的实时额度。")
        }
        guard !windows.isEmpty, windows.contains(where: { $0.remaining != nil }) else {
            throw MonitorError("未返回可识别的额度；不会按 0% 处理。")
        }
        return QuotaResult(windows: windows, plan: plan)
    }
}

public struct AlertLedger: Codable {
    public struct Entry: Codable {
        var reset: Date?
        var thresholds: Set<Int>
        var touched: Date
    }
    public var entries: [String: Entry] = [:]
    public init() {}
    public mutating func evaluate(key: String, window: QuotaWindow, warning: Int, critical: Int,
                                  now: Date, maxAge: TimeInterval) -> Int? {
        guard window.fresh(now: now, maxAge: maxAge), let remaining = window.remaining else { return nil }
        var entry = entries[key] ?? Entry(reset: window.reset, thresholds: [], touched: now)
        // Providers can move reset timestamps by a few seconds within the same cycle.
        if let old = entry.reset, let new = window.reset, new.timeIntervalSince(old) > 120, now >= old {
            entry.thresholds = []
        } else if entry.reset == nil && window.reset == nil && remaining > Double(warning + 5) {
            entry.thresholds = []
        }
        entry.reset = window.reset
        entry.touched = now
        let threshold = remaining <= Double(critical) ? critical : remaining <= Double(warning) ? warning : nil
        let shouldAlert = threshold.map { !entry.thresholds.contains($0) } ?? false
        if let threshold, shouldAlert {
            entry.thresholds.insert(threshold)
            if threshold == critical { entry.thresholds.insert(warning) }
        }
        entries[key] = entry
        entries = entries.filter { now.timeIntervalSince($0.value.touched) < 90 * 86400 }
        return shouldAlert ? threshold : nil
    }
}

File: windows/MonitorCore/Domain.cs
using System.Globalization;
using System.Text.Json.Nodes;

namespace CPAMP.MonitorCore;

public sealed class MonitorException(string message, int? status = null) : Exception(message)
{
    public int? Status { get; } = status;
}

public sealed record Configuration
{
    public string BaseUrl { get; init; } = "";
    public string Name { get; init; } = "CPAMP";
    public int Interval { get; init; } = 300;
    public int Warning { get; init; } = 20;
    public int Critical { get; init; } = 10;
    public bool Notifications { get; init; } = true;
    public HashSet<string> Excluded { get; init; } = [];
    public List<string> AccountOrder { get; init; } = [];
    public double MaxAge => Interval * 2 + 60;

    public static Uri NormalizeUrl(string input)
    {
        if (!Uri.TryCreate(input.Trim(), UriKind.Absolute, out var uri) ||
            uri.Scheme is not ("http" or "https") || string.IsNullOrEmpty(uri.Host) ||
            uri.UserInfo.Length != 0 || uri.Query.Length != 0 || uri.Fragment.Length != 0)
            throw new MonitorException("Enter a service URL without credentials, query parameters or fragments.");
        if (uri.Scheme == "http" && uri.Host.ToLowerInvariant() is not ("localhost" or "127.0.0.1" or "[::1]"))
            throw new MonitorException("Remote services must use HTTPS.");
        var builder = new UriBuilder(uri);
        var path = uri.AbsolutePath.TrimEnd('/');
        foreach (var suffix in new[] { "/management.html", "/v0/management", "/v1" })
            if (path.EndsWith(suffix, StringComparison.Ordinal)) { path = path[..^suffix.Length]; break; }
        builder.Path = path + "/";
        return builder.Uri;
    }

    public Configuration Validate()
    {
        if (Interval is < 60 or > 3600 || Critical < 1 || Warning > 99 || Critical >= Warning)
            throw new MonitorException("Refresh: 1-60 minutes. Thresholds: 1 <= critical < warning <= 99.");
        return this with { BaseUrl = NormalizeUrl(BaseUrl).AbsoluteUri, Name = Name.Trim() };
    }
}

public static class Json
{
    public static string Text(JsonNode? node) => node is JsonValue v ? v.ToString() : "";
    public static double? Number(JsonNode? node)
    {
        if (node is not JsonValue v || !double.TryParse(v.ToString(), NumberStyles.Float,
                CultureInfo.InvariantCulture, out var n) || !double.IsFinite(n)) return null;
        return n;
    }
    public static double? Percent(JsonNode? node) => Number(node) is >= 0 and <= 100 ? Number(node) : null;
    public static DateTimeOffset? Timestamp(JsonNode? node)
    {
        if (Number(node) is > 0 and < 100_000_000_000_000 and var n)
        {
            try { return DateTimeOffset.UnixEpoch.AddSeconds(n > 100_000_000_000 ? n / 1000 : n); }
            catch (ArgumentOutOfRangeException) { return null; }
        }
        return DateTimeOffset.TryParse(Text(node), CultureInfo.InvariantCulture,
            DateTimeStyles.AssumeUniversal, out var date) ? date : null;
    }
    public static IEnumerable<JsonObject> Objects(JsonNode? node) =>
        node is JsonArray a ? a.OfType<JsonObject>() : [];
}

public sealed record Account(string Filename, string Provider, string AuthIndex, string Email,
    string Status, bool Disabled, string ProjectId, string AccountId, string Plan)
{
    public string Id => $"{Provider}:{Filename}:{AuthIndex}";
    public string Title => string.IsNullOrEmpty(Email) ? Filename : Email;
    public string ProviderName => Provider switch { "codex" => "Codex", "xai" => "Grok / xAI",
        "antigravity" => "Antigravity", "claude" => "Claude", _ => Provider };
    public static Account FromJson(JsonObject j)
    {
        var a = j["account"] as JsonObject;
        return new(Json.Text(j["name"]), Json.Text(j["provider"] ?? j["type"]).ToLowerInvariant(),
            Json.Text(j["auth_index"]), Json.Text(j["email"]), Json.Text(j["status"]),
            j["disabled"] is JsonValue d && d.TryGetValue<bool>(out var disabled) && disabled,
            Json.Text(j["project_id"]), Json.Text(a?["account_id"] ?? j["account_id"]),
            Json.Text(a?["plan_type"] ?? j["account_type"]));
    }
    public JsonObject Target()
    {
        var target = new JsonObject { ["row_key"] = Id, ["auth_file_snapshot"] = Filename,
            ["auth_provider_snapshot"] = Provider, ["auth_index"] = AuthIndex };
        if (AccountId.Length > 0) target["auth_account_id_snapshot"] = AccountId;
        if (ProjectId.Length > 0) target["auth_project_id_snapshot"] = ProjectId;
        return target;
    }
}

public sealed record QuotaWindow(string Id, string Title, double? Remaining, DateTimeOffset? Reset,
    DateTimeOffset Observed)
{
    public bool Fresh(DateTimeOffset now, double maxAge) => (now - Observed).TotalSeconds >= -60 &&
        (now - Observed).TotalSeconds <= maxAge && (Reset is null || Reset > now);
}
public sealed record QuotaResult(List<QuotaWindow> Windows, string Plan = "");

public static class QuotaParser
{
    public static QuotaResult Parse(string provider, JsonObject payload, DateTimeOffset now)
    {
        List<QuotaWindow> windows = [];
        var plan = "";
        void Add(string id, string title, double? remaining, DateTimeOffset? reset) =>
            windows.Add(new(id, title, remaining, reset, now));
        switch (provider)
        {
            case "codex":
                plan = Json.Text(payload["plan_type"]);
                List<(string Id, string Prefix, JsonObject Limit)> limits = [];
                if (payload["rate_limit"] is JsonObject main) limits.Add(("main", "", main));
                if (payload["code_review_rate_limit"] is JsonObject review) limits.Add(("review", "Code review ", review));
                var extraIndex = 0;
                foreach (var extra in Json.Objects(payload["additional_rate_limits"]))
                {
                    var name = Json.Text(extra["limit_name"] ?? extra["metered_feature"]);
                    if (extra["rate_limit"] is JsonObject rate)
                        limits.Add(("extra-" + (name.Length == 0 ? extraIndex.ToString() : name), name + " ", rate));
                    extraIndex++;
                }
                foreach (var (id, prefix, limit) in limits)
                    foreach (var key in new[] { "primary_window", "secondary_window" })
                    {
                        if (limit[key] is not JsonObject w) continue;
                        var duration = Json.Number(w["limit_window_seconds"]);
                        var label = duration == 604800 ? "Weekly" : duration.HasValue ? $"{duration / 3600:0} hours" : "Quota";
                        var reset = Json.Timestamp(w["reset_at"]);
                        if (reset is null && Json.Number(w["reset_after_seconds"]) is > 0 and < 315360000 and var seconds)
                            reset = now.AddSeconds(seconds);
                        Add(id + ":" + key, prefix + label, 100 - Json.Percent(w["used_percent"]), reset);
                    }
                break;
            case "antigravity":
                var gi = 0;
                foreach (var group in Json.Objects(payload["groups"]))
                {
                    var label = Json.Text(group["displayName"] ?? group["display_name"]);
                    var bi = 0;
                    foreach (var bucket in Json.Objects(group["buckets"]))
                    {
                        var period = Json.Text(bucket["window"]);
                        var fraction = Json.Number(bucket["remainingFraction"] ?? bucket["remaining_fraction"]);
                        var id = Json.Text(bucket["bucketId"] ?? bucket["bucket_id"]);
                        Add(id.Length == 0 ? $"{label}-{gi}-{period}-{bi}" : id,
                            $"{(label.Length == 0 ? $"Quota group {gi + 1}" : label)} {(period == "weekly" ? "Weekly" : period)}",
                            fraction is >= 0 and <= 1 ? fraction * 100 : null,
                            Json.Timestamp(bucket["resetTime"] ?? bucket["reset_time"]));
                        bi++;
                    }
                    gi++;
                }
                break;
            case "xai":
                var config = payload["config"] as JsonObject;
                var current = config?["currentPeriod"] as JsonObject;
                if (Json.Percent(config?["creditUsagePercent"]) is double used)
                    Add("billing", Json.Text(current?["type"]).Contains("WEEKLY") ? "Weekly" : "Billing period",
                        100 - used, Json.Timestamp(current?["end"] ?? config?["billingPeriodEnd"]));
                break;
            case "claude":
                foreach (var (key, value) in payload.OrderBy(p => p.Key))
                    if ((key.StartsWith("five_hour") || key.StartsWith("seven_day")) && value is JsonObject w)
                        Add(key, key == "five_hour" ? "5 hours" : key == "seven_day" ? "Weekly" : key,
                            100 - Json.Percent(w["utilization"]), Json.Timestamp(w["resets_at"]));
                break;
            default: throw new MonitorException("Live quota is not supported for this provider.");
        }
        if (!windows.Any(w => w.Remaining.HasValue))
            throw new MonitorException("No recognized quota returned. Unknown values are not treated as zero.");
        return new(windows, plan);
    }
}

public sealed class AlertEntry
{
    public DateTimeOffset? Reset { get; set; }
    public HashSet<int> Thresholds { get; set; } = [];
    public DateTimeOffset Touched { get; set; }
}
public sealed class AlertLedger
{
    public Dictionary<string, AlertEntry> Entries { get; set; } = [];
    public int? Evaluate(string key, QuotaWindow window, int warning, int critical,
        DateTimeOffset now, double maxAge)
    {
        if (!window.Fresh(now, maxAge) || window.Remaining is not double remaining) return null;
        if (!Entries.TryGetValue(key, out var entry)) entry = new() { Reset = window.Reset };
        if (entry.Reset is DateTimeOffset old && window.Reset is DateTimeOffset next &&
            (next - old).TotalSeconds > 120 && now >= old) entry.Thresholds.Clear();
        else if (entry.Reset is null && window.Reset is null && remaining > warning + 5) entry.Thresholds.Clear();
        entry.Reset = window.Reset;
        entry.Touched = now;
        int? threshold = remaining <= critical ? critical : remaining <= warning ? warning : null;
        var alert = threshold.HasValue && entry.Thresholds.Add(threshold.Value);
        if (alert && threshold == critical) entry.Thresholds.Add(warning);
        Entries[key] = entry;
        foreach (var stale in Entries.Where(p => (now - p.Value.Touched).TotalDays >= 90).Select(p => p.Key).ToArray())
            Entries.Remove(stale);
        return alert ? threshold : null;
    }
}
